src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -4345,39 +4407,41 @@ private void sendClose() throws Exception
 		}
 	}
 
-	public int sendParameterMessage(final ClientWireProtocol.SetParameter param) throws SQLException
+	public CompletionStage<SQLResult<Integer>> sendParameterMessage(final ClientWireProtocol.SetParameter param)
 	{
 		final ClientWireProtocol.Request.Builder builder = ClientWireProtocol.Request.newBuilder();
 		builder.setType(ClientWireProtocol.Request.RequestType.SET_PARAMETER);
 		builder.setSetParameter(param);
 		final Request wrapper = builder.build();
 
-		try
-		{
-			// FIXME make async
-			blockingGetSQLResult(sendRequestWithStandardResponse(
-								wrapper, 
-								Functions::alwaysTrue)).get();
-		} catch (final Exception ex)
-		{
-			if(ex instanceof SQLException && SQLStates.SESSION_EXPIRED.equals((SQLException) ex)){
-				LOGGER.log(Level.INFO, "sendParameterMessage() received session expired. Attempting to refresh session");
-				// Refresh the now-expired session.
-				// FIXME this is now async
-				try {
-					blockingGet(refreshSession().toCompletableFuture());
+		// Return zero on success
+		return sendRequestWithStandardResponse(wrapper, Functions::alwaysTrue)
+			.thenCompose(respOrEx -> respOrEx.recoverAsync(ex -> {
+
+				if (ex.getErrorCode() == SQLStates.SESSION_EXPIRED.getSqlCode()) {
+					// Refresh session and try again
+					return sendRefreshSession()
+						.thenCompose(successOrEx -> successOrEx
+						.mapSuccessAsync(ign -> sendRequestWithStandardResponse(wrapper, Functions::alwaysTrue)
+							.thenApply(retryOrEx -> {
+
+								if(retryOrEx.isException()) {
+									// Retry (after refreshing the session) failed
+									SQLException e = retryOrEx.getException();
+									LOGGER.log(Level.WARNING, e, () -> String.format("Failed sending set parameter request to the server after refresh with exception %s with message %s", e, e.getMessage()));
```

Comment:
(nit)

Avoid printing the message which will shown up in the stack trace and include `this` which has a `toString()` method that will let us know which channel the error happened on.
```suggestion
									LOGGER.log(Level.WARNING, e, () -> String.format("Failed sending set parameter request to the server after refresh: %s", this));
```

There's more I want to do here w.r.t. contextual logging. There's  pattern referred to as `MDC` in logging utilities that allows you to define thread local variables to provide additional context to log messages implicitly i.e. a query ID used to prefix each message. The pattern is trickier to implement in a non-blocking world where code can run an any thread so I didn't do that, be we should consider doing something like that in the future.